### PR TITLE
Feature/protocol compatible tombstones

### PR DESCRIPTION
**Pull request by thanodnl**
**IMPORTANT:** This diff is against a feature branch who's sole purpose it is to revert the revert of the reaping-faulty-nodes feature. When this lands an other PR will be opened as a formality to get all changes to `dev`.

These changes implement a new protocol for gossipping tombstones that is compatible with previous versions on top of the changes revert the revert of the reaping faulty nodes feature.

Main things to note here is that tombstones are gossiped around as `faulty` state with a `tombstone` flag on the change. This flag is ignored by older versions of ringpop and treated as the faulty state. This make sure that the `panic` condition is not triggered while gossiping tombstones during a hybrid deployment of the old and the new version.

Next there is an admin endpoint that can be used to manually initiate the reaping of faulty nodes. This will also be implemented in `ringpop-admin` for users to call after a successful upgrade to validate that the reaping works without waiting the _faulty period_.

Both changes have accompanying integration tests added to the `reaping-faulty-nodes` feature. This will be merged to the master branch of `ringpop-common` right before these changes land on `dev`.

When older versions of ringpop participate in the membership of tombstone aware nodes these changes will prevent the older versions from crashing on a panic parsing the tombstone state. Instead the tombstone flag will be ignored. This causes checksum incompatibilities during the _tombstone period_ between the older versions and the newer version. This has a side effect of elevated full syncs happening during this period which in the end causes removed members to reappear and be redeclared a tombstone after the _faulty period_. Only when all running instances run a tombstone aware version the  faulty's can be reaped effectively.

At the same time the panic statement that caused crashes is removed from the precedence calculation of states. This prevents hard crashes in the future.

### CODE REVIEW

**Comment by thanodnl** [comment]
@Motiejus @severb Can you have a look at this PR? It is the PR discussed in the Standup that adds wire-protocol compatibility with older ringpop's for the tombstone changes.

File: swim/member.go [hunk 1]
```diff
@@ -128,11 +129,27 @@ type Change struct {
 	Address           string `json:"address"`
 	Incarnation       int64  `json:"incarnationNumber"`
 	Status            string `json:"status"`
+	Tombstone         bool   `json:"tombstone,omitempty"`
 	// Use util.Timestamp for bi-direction binding to time encoded as
 	// integer Unix timestamp in JSON
 	Timestamp util.Timestamp `json:"timestamp"`
 }
 
+func (c Change) validateIncomming() Change {
```
Comment (Motiejus) [inline swim/member.go]: s/Incomming/Incoming/

File: swim/memberlist.go [hunk 2]
```diff
@@ -304,6 +304,11 @@ func (m *memberlist) Update(changes []Change) (applied []Change) {
 		return nil
 	}
 
+	// validate incomming changes
```
Comment (Motiejus) [inline swim/memberlist.go]: s/incomming/incoming

**Comment by Motiejus** [comment]
Sorry for taking that long to look at this.

Given our current situation in terms of compatibility, I really like your approach.

**Comment by Motiejus** [comment]
LGTM

File: test/run-integration-tests [hunk 3]
```diff
@@ -6,7 +6,7 @@ set -eo pipefail
 
 declare project_root="${0%/*}/.."
 declare ringpop_common_dir="${0%/*}/ringpop-common"
-declare ringpop_common_branch="master"
+declare ringpop_common_branch="feature/admin-reap"
```
Comment (severb) [inline test/run-integration-tests]: I don't think we should push this change to ringpop-go master.

**Comment by severb** [comment]
LGTM

File: test/run-integration-tests (hunk 3 above)
Comment (thanodnl) [inline test/run-integration-tests]: I completely agree. That branch needs to land to master in `ringpop-common` before we merge this branch.

**Comment by severb** [comment]
Awesome job! 🎩

**Comment by dansimau** [comment]
winner

**Pull request merged**